Libraries/Opc.Ua.Client/Session.cs

```diff
@@ -6059,18 +6052,19 @@ private CallMethodRequestCollection CreateCallRequestsForResendData(IEnumerable<
         /// <summary>
         /// Create the publish requests for the active subscriptions.
         /// </summary>
-        private void RestartPublishing()
+        public void StartPublish(int timeout)
         {
-            int publishCount = 0;
-            lock (SyncRoot)
+            int publishCount = GetMinPublishRequestCount(true);
+
+            if (m_tooManyPublishRequests > 0 && publishCount > m_tooManyPublishRequests)
             {
-                publishCount = GetMinPublishRequestCount(true);
+                publishCount = m_tooManyPublishRequests;
             }
 
             // refill pipeline.
-            for (int ii = 0; ii < publishCount; ii++)
+            for (int ii = GoodPublishRequestCount; ii < publishCount; ii++)
             {
```

Comment:
GoodPublishRequestCount is using a lock and walks a list. Not sure if its always needed. At minimum, one BeginPublish should be sent.